zerver/actions/message_send.py

```diff
@@ -210,11 +214,23 @@ def get_recipient_info(
         # of this function for different message types.
         assert stream_topic is not None
 
+        topic_participant_user_ids: AbstractSet[int] = set()
+        if possible_topic_wildcard_mention:
+            # A topic participant is anyone who sent a message to the topic or
+            # reacted to a message on the topic.
+            messages = messages_for_topic(recipient.id, stream_topic.topic_name)
+            topic_participant_user_ids = ({message.sender_id for message in messages}) | set(
+                Reaction.objects.filter(message__in=messages).values_list(
+                    "user_profile_id", flat=True
+                )
+            )
```

Comment:
This should probably be turned into a `zerver.lib.topic` method `participants_for_topic` -- it's a clear enough definition that we may find we want to use it for something else later, and I think it'll help readability. 

We should also document that function as potentially expensive -- the `Message` part is pretty cheap, but the `Reaction` part may or may not be if there's a lot of messages in the topic. And add a comment here that it's OK to do this potentially expensive thing since we only do it when `@topic` appears in the message body.

(It's also the sort of thing that could be optimized to a single query if profiling suggested that's useful).

Reply:
[#25828 Comment](https://github.com/zulip/zulip/pull/25828#discussion_r1240792510)